### PR TITLE
Support for managing EC2 Volumes

### PR DESCRIPTION
**Pull request by daveseff**
I am adding support for managing EC2 Block device volumes. With this you can create and attach volumes to already running instances as well as detach and destroy. 

Volumes can be created from scratch or built from a snapshot. 

TODO: 
- Add the ability to reference ec2_instances by name. I found that doing ec2_instance name lookups is really slow so for performance I left it out. If someone knows a better way, please let me know.

### CODE REVIEW

File: spec/unit/provider/ec2_volume/v2_spec.rb [hunk 1]
```diff
@@ -0,0 +1,29 @@
+require 'spec_helper'
+
+provider_class = Puppet::Type.type(:ec2_volume).provider(:v2)
+
+ENV['AWS_ACCESS_KEY_ID'] = 'redacted'
+ENV['AWS_SECRET_ACCESS_KEY'] = 'redacted'
+ENV['AWS_REGION'] = 'sa-east-1'
```
Comment (DavidS) [inline spec/unit/provider/ec2_volume/v2_spec.rb]: This will corrupt the environment for all other test in the suite. Please remove.

**Comment by DavidS** [comment]
This looks quite solid.

The only thing really missing is a acceptance test that exercises the type and provider. See https://github.com/puppetlabs/puppetlabs-aws/blob/master/spec/acceptance/loadbalancer_spec.rb for a simple example.

Regarding the TODOs, I think encrypted volumes are not a necessary requirement to get this merged. It is useful as it is. The id instead of name to reference instances is more of an issue as the rest of the module currently allows users to always use the name, and the id is chosen by AWS, and not the user. One possibility would be to create a cache (bound to the lifetime of a puppet agent transaction) that maps between instance IDs and names. This could come in handy in a number of places.

**Comment by zleswomp** [comment]
I'd love to see this make it in.  Are tests in progress here, or is something that is waiting for another community member to pick up?

**Comment by daveseff** [comment]
I've been working on it when I have time to put into it. My company has a need for this feature which is why I have added it. It's working great for us and I have had to move onto other fires. I want to get this merged in soon so I'll have to figure out how to complete the tests.

**Comment by zleswomp** [comment]
@daveseff Hows this going?  I'm keen to see this make it in, as we have this need also.

**Comment by daveseff** [comment]
I thought I had all of the tests done. I don't see why this couldn't get merged as my company has been using this feature for several months now.

File: lib/puppet/provider/ec2_volume/v2.rb [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+    config[:volume_id] = volume_id
+    config[:device] = resource[:attach]["device"]
+    Puppet.info("Attaching Volume #{volume_id} to ec2 instance #{config[:instance_id]}")
+    ec2.wait_until(:volume_available, volume_ids: [volume_id])
+    ec2.attach_volume(config)
+  end
+
+  def create
+    Puppet.info("Creating Volume #{name} in region #{target_region}")
+    #ec2 = ec2_client(resource[:region])
```
Comment (hunner) [inline lib/puppet/provider/ec2_volume/v2.rb]: This isn't needed, yeah?

**Comment by hunner** [comment]
Just a few comments. Could you squash it down to one commit? Thanks!

File: README.md [hunk 3]
```diff
@@ -528,6 +529,35 @@ back- end instances.  Accepts a hash with the following keys:
 #####`scheme`
 *Optional* Whether the load balancer is internal or public facing. This parameter is set at creation only; it is not affected by updates. Valid values are 'internal', 'internet-facing'. Default value is 'internet-facing' and makes the load balancer publicly available.
 
+#### Type: ec2_volume
+
+##### `name`
+*Required* The name of the volume
+EEEEEEEEE
```
Comment (daveseff) [inline README.md]: derp!

**Comment by daveseff** [comment]
I had it as one commit. Seems my rebase against master has added a bunch of stuff. I'll try again

**Comment by daveseff** [comment]
New PR created. (#400)

**Pull request closed without merging**